### DISCUSSION TITLE
Implementing AbstractCanvas2D using canvas

### DISCUSSION THREAD

**Discussion by cd-yang** [maxGraph/maxGraph#182] [Development]
This discussion is intended for discussing the process of implementing AbstractCanvas2D using canvas. 
#180 
As a start, I have implemented the drawing of blocks and lines using canvas, as shown in the image below.
<img width="1262" alt="image" src="https://user-images.githubusercontent.com/17699525/222963751-23b92d04-4c49-4f31-b9d3-7d412b7cd620.png">

I would like to share some of the issues I encountered during the coding process:

- according to the Least Knowledge Principle, many methods and properties in AbstractCanvas2D and SvgCanvas2D can be set to private.
- the static method `createCss` can be removed from SvgCanvas2D.
- implementing the rendering method is relatively easy, but the challenge lies in handling mouse click events. It may be necessary to maintain a separate mechanism for the position of the cells (maybe similar to GraphView) to properly handle mouse click events.

**Comment by cd-yang** [comment]
I'm using `console.log` for debugging temporarily.
Are there any simple ways to add breakpoints in VS Code. Do I need to create a new project and directly reference "packages/core" for debugging with breakpoints?

**Reply by tbouffard** [reply]
Hi and sorry for the late answer 😸 
Did you find a solution to your problem?

The regular VSCode debugger should work https://code.visualstudio.com/docs/typescript/typescript-debugging
However, we may have an issue with the way the maxGraph code is built and used in the storybook example.
Storybook uses the `maxGraph` UMD bundle built with `webpack`. The sourcemap may not be generated at all or may be wrong. It is required to map the JS code within the UMD bundle with the TypeScript sources: this is a prerequisites to make the debugger work.
If the sourcemap is correctly configured, it should work in VSCode and in browser DevTools as well.

**Reply by cd-yang** [reply]
Thanks. I'll try it next week.

**Comment by cd-yang** [comment]
#187 
I just created a PR to demonstrate my current progress. Last week, I tried to implement the selection feature for shapes on the canvas, but encountered some difficulties and did not complete it.

**Reply by tbouffard** [reply]
Thanks for sharing.
I don't have time to have a look at it for now but I will give it a try as soon as possible.

**Comment by KissBalazs** [comment]
Are there any further progress with this initiative? We are currently searching for tools for one of our projects with my team, and maxGraph would be the absolute best fit, if it would support canvas (for performance reasons mostly).
